Added schema validation for policies (WP-27)

smonteleone opened the pull request:
Fixed default policy and some jasmine tests

Jira issue: WP-27

smonteleone committed:
Added schema validation for policies (WP-27)
Fixed default policy and some jasmine tests

Jira issue: WP-27

habibvirji commented on webinos/core/manager/policy_manager/lib/policymanager.js:
```diff
@@ -24,6 +24,12 @@
 	var promptLib = require('../src/promptMan/promptManager.js');
 	var dslib = require('./decisionstorage.js');
+    var JSV = require('JSV').JSV;
+    var fs = require('fs');
```
indentation is different

habibvirji commented on webinos/core/manager/policy_manager/lib/policymanager.js:
```diff
@@ -31,8 +37,9 @@
+    var json = null;
 
-	var policyManager = function(policyFilename) {
+    var policyManager = function(policyFilename) {
```
If you use this here, it will be this of not policy manager but of the entity calling it. Should suggest if you intend to use single instance then change this to refer to policy manager.  This will create separate instance of each call. 

I usually solve this by declaring self = this in the main function and then use self instead of this.

smonteleone replied:
Thank you for the suggestion.. Fixed with commit 5909889

smonteleone committed:
Fixed policymanager.js
fixed reference and small indentation issues

Jira issue: WP-27

jplyle commented:
I get weird errors in Linux.

If I start the PZP, then stop the PZP, then go and fiddle with the policy.xml file (intentionally making it wrong) I get:

New packet from messaging
Response to deviceA_Pzp/0
[ReferenceError: self is not defined]
ReferenceError: self is not defined
    at policyManager.enforceRequest (/home/johl/git-repos/Webinos-Platform/webinos/core/manager/policy_manager/lib/policymanager.js:77:13)
    at handleMessage [as checkPolicy](/home/johl/git-repos/Webinos-Platform/webinos/core/manager/policy_manager/lib/rpcInterception.js:72:16)
    at _RPCHandler.handleMessage (/home/johl/git-repos/Webinos-Platform/node_modules/webinos-jsonrpc2/lib/rpc.js:303:25)
    at MessageHandler.onMessageReceived (/home/johl/git-repos/Webinos-Platform/webinos/core/manager/messaging/lib/messagehandler.js:400:23)
    at wsMessage (/home/johl/git-repos/Webinos-Platform/webinos/core/pzp/lib/pzp_websocket.js:186:51)
    at WebSocketConnection.PzpWSS.startWebSocketServer (/home/johl/git-repos/Webinos-Platform/webinos/core/pzp/lib/pzp_websocket.js:651:72)
    at WebSocketConnection.EventEmitter.emit (events.js:96:17)
    at WebSocketConnection.processFrame (/home/johl/git-repos/Webinos-Platform/node_modules/websocket/lib/WebSocketConnection.js:357:26)
    at WebSocketConnection.handleSocketData (/home/johl/git-repos/Webinos-Platform/node_modules/websocket/lib/WebSocketConnection.js:206:14)
    at Socket.EventEmitter.emit (events.js:96:17)
[22.1.2013-14:35:56:48] info pzp_websocket.js(678) [deviceA_Pzp]  send to deviceA_Pzp/0 message {"type":"prop","from":"deviceA_Pzp","to":"deviceA_Pzp/0","payload":{"status":"webinosVersion","message":"v0.8.0-8-g5909889\n"}}

jplyle commented:
Actually, I'm now getting that when I haven't changed the policy.xml file.

smonteleone committed:
Removed wrong reads of "self" properties

Jira issue: WP-27

smonteleone commented:
It works for me now (Tested on linux and OSX)

jplyle commented:
Could you rebase?  Then I'll do another test.  Sorry for the slow reply...

smonteleone committed:
Merge branch 'master' into WP-27

Jira issue: WP-27

smonteleone committed:
Fixed xml2js parsing issue.

Jira issue: WP-27

jplyle commented:
I'm now very happy with this patch - it works beautifully on Linux.  All policy manager tests pass, too... except i'm getting a segfault toward the end.  Will investigate!

jplyle commented:
johl@ubuntu:~/git-repos/Webinos-Platform/webinos/core/manager/policy_manager/test/jasmine$ 
jasmine-node .

Change policy to file policy_dhp_5.xml
Decision storage constructor
Permanent file is decisionpermanent.xml
Policy editor constructor
Policy file loaded
Setting request for user user3, device device2, application released by Company2, feature http://webinos.org/api/messaging.send, purpose true,false,false,false,true,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false and obligations undefined
Policy Manager enforce request: {"subjectInfo":{"userId":"user3"},"widgetInfo":{"distributorKeyCn":"Company2"},"deviceInfo":{"requestorId":"device2"},"resourceInfo":{"apiFeature":"http://webinos.org/api/messaging.send"},"purpose":[true,false,false,false,true,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false]} - result is 1
result is 1
.Change policy to file policy_dhp_6.xml
Segmentation fault (core dumped)

smonteleone committed:
Changed policy tags to match the grammar
(Uri->AccessURI, Start->StartTime)
Restored Purpose value for policy_dhp_12.xml

Jira issue: WP-27

smonteleone committed:
Merge branch 'master' into WP-27

smonteleone committed:
Changed default policy
With this one, policy manager allows any request (access control +
privacy)

Jira issue: WP-27

jplyle commented:
I'm not getting any errors with this.

I am, however, getting some odd errors with the PZH.  This may not be related to your pull request...

jplyle commented:
... having compared with master, there seems to be a problem.

When you log into the PZH in your branch, the PZH shows the wrong 'connected personal zone' .  

Any ideas?

smonteleone commented:
This is what I see on my mac + virtual linux setup.. 
I think there was a problem merging older versions (as you can see I have a "Webinos Version - v0.5.1366( CommitId-gd0e0ab3)" label on top-right corner) but also APIs work..

![screenshot_linux_pzh](https://f.cloud.github.com/assets/1619042/224470/d5c5ff84-85d5-11e2-93da-75ba8cb60dc6.png)
![screenshot_osx](https://f.cloud.github.com/assets/1619042/224433/cfaed89c-85d4-11e2-9a8a-2fa914ee2c3a.png)
![screenshot_linux](https://f.cloud.github.com/assets/1619042/224406/4915d808-85d4-11e2-816f-3ff9465c62d1.png)

smonteleone commented:
I've  found out that  the problem you mention happens also on my phone with the apk (version 0.8 downloaded from webinos site) using pzh.webinos.org

![Screenshot_2013-03-06-02-26-19](https://f.cloud.github.com/assets/1619042/225593/30a59a8c-85fd-11e2-89e4-d08d3d4b7818.png)

smonteleone commented:
The same happens on my laptop (with the last master from http://github.com/webinos/Webinos-Platform.git).. In both cases my account is recognised as "Bob Mobilis (google)" account.

![Screen Shot 2013-03-06 at 2 58 53 AM](https://f.cloud.github.com/assets/1619042/225690/2554a448-8602-11e2-9c5c-4fb8035e710b.png)

jplyle commented:
This is something that @habibvirji should be aware of...
